### PR TITLE
Fix JSON, YAML `use_*_discriminator` for recursive struct types

### LINKED ISSUE #13214: JSON::Serializable::Strict compile bug
this code failed to compile, which is pretty weird, because if I remove `JSON::Serializable::Strict` it compiled. But Strict just add one exception, it should not behave like this.

```crystal
require "json"

class Base
  include JSON::Serializable
  include JSON::Serializable::Strict

  use_json_discriminator("type", {
    "a": A,
    "b": B,
  })

  property type : String
end

class A < Base
end

class B < Base
  property child1 : Base
  property child2 : Base
end

Base.from_json("{\"type\": \"b\", \"child1\": {\"type\": \"a\"}, \"child2\": {\"type\": \"a\"}}")
```

https://play.crystal-lang.org/#/r/er0s
```
Showing last frame. Use --error-trace for full trace.

There was a problem expanding macro 'macro_4604816000'

Code in /opt/homebrew/Cellar/crystal/1.7.3/share/crystal/src/json/serialization.cr:182:7

 182 | {% begin %}
       ^
Called macro defined in /opt/homebrew/Cellar/crystal/1.7.3/share/crystal/src/json/serialization.cr:182:7

 182 | {% begin %}

Which expanded to:

 > 135 |
 > 136 |
 > 137 |             @child2 = (__temp_966).as(Base)
                               ^
Error: can't cast Nil to Base
```

### PR DESCRIPTION
**Pull request by HertzDevil**
Fixes #13214. This makes sure the temporary variables are not nilable for properties that aren't nilable and have no default values, since otherwise it seems the deserialization method recursively calls itself and breaks type inference.

This was discovered for JSON but applies to YAML as well.

### CODE REVIEW

File: src/json/serialization.cr [hunk 1]
```diff
@@ -216,26 +216,24 @@ module JSON
...
-                  {% if value[:root] %}
-                    pull.on_key!({{value[:root]}}) do
-                  {% end %}
-
-                  {% if value[:converter] %}
-                    {{value[:converter]}}.from_json(pull)
-                  {% else %}
-                    ::Union({{value[:type]}}).new(pull)
-                  {% end %}
+                {% if value[:has_default] || value[:nilable] %} pull.read_null_or do {% else %} begin {% end %}
```
Comment (straight-shoota) [inline src/json/serialization.cr:220]: Is `begin` actually necessary here or is it just the counterpart for an unconditional `end` later?

**Review by straight-shoota — APPROVED** [review]

File: src/json/serialization.cr [hunk 2]
```diff
@@ -200,7 +200,7 @@ module JSON
         {% end %}
 
         {% for name, value in properties %}
-          %var{name} = nil
+          %var{name} = {% if value[:has_default] || value[:nilable] %} nil {% else %} uninitialized {{value[:type]}} {% end %}
```
Comment (kostya) [inline src/json/serialization.cr:203]: probably needs wrap: `uninitialized ::Union({{value[:type]}})`

File: src/json/serialization.cr (hunk 1 above)
Comment (HertzDevil) [reply src/json/serialization.cr:220]: The existing code guarded both the `{` and the `}`, but I think the code indents better this way.

**Comment by straight-shoota** [comment]
I think the proper action is to raise right away instead of returning `nil`.

Suppose the deserializer could technically type the variable as `typeof(::Union({{value[:type]}}).new(pull))` and verify the actual runtime value at the end. This would preserve this use case while still fixing the invalid nil propagation issue.

**Comment by straight-shoota** [comment]
It's unfortunate to see this previously working code break. But I think it's for good. It only worked due to a bug in the first place and as far as I can tell was never intended nor documented behaviour.

We could re-enable that behaviour by adding in an appropriate cast in the deserialization logic. But I don't like that. It trades away type safety for no real benefit out of itself only to keep existing code that relied on a bug working.
This is a tough call of course. Should we prioritize avoiding to break code that worked previously out of error? We value backwards compatibility, but in this situation it depends on goodwill.

I believe it's best to accept the breaking code. The quality of the library will be better that way. And as the last example shows (https://github.com/crystal-lang/crystal/pull/13238#issuecomment-1512096511) the breakage even helps to discover type errors.
That's the point of having type safety. Giving that away opens up other holes.

It's easy to fix the affected code by raising directly in the converter. This is also backwards compatible with older Crystal releases.

**Pull request merged**